### PR TITLE
Add documentation

### PR DESCRIPTION
**Pull request by rmfan**

### CODE REVIEW

File: src/graph/controller.py [hunk 1]
```diff
@@ -43,7 +51,12 @@ def __init__(self):
         self._deployment_lock = asyncio.Lock()
 
     async def run(self):
-        # Create service if it doesn't already exist.
+        """
+        Spawn non-terminating tasks to watch changes to RampartGraph CRD objects and update
```
Comment (DianLiI) [inline src/graph/controller.py]: It should CR objects not CRD objects, same to the line below.
I've added a comment for each of such appearances. But to make our lives easier I think we can delete all the CRDs if there is already a RampartGraph before it.

File: src/graph/controller.py [hunk 2]
```diff
@@ -52,6 +65,15 @@ async def run(self):
         )
 
     async def _patch_graph_status(self, namespace, name, patch_obj):
+        """
+        Patch the RampartGraph CRD with the new values in patch_obj.
```
Comment (DianLiI) [inline src/graph/controller.py]: CR. Same to the lines under Args.

File: src/graph/controller.py [hunk 3]
```diff
@@ -65,6 +87,11 @@ async def _patch_graph_status(self, namespace, name, patch_obj):
         raise exception
 
     async def _watch_graphs(self):
+        """
+        A non-terminating function that watches changes to RampartGraph CRD's across
```
Comment (DianLiI) [inline src/graph/controller.py]: All appearances of CRDs should be CR in this comment

File: src/graph/controller.py [hunk 4]
```diff
@@ -97,22 +129,42 @@ async def _sync_graphs(self):
             self._queue.task_done()
 
     async def _sync_graph(self, metadata):
+        """
+        This function wraps the tasks that do the actual work of updating the cluster
+        to manage waiting for completion and cancellation.
+
+        args:
+            metadata (objects.base_type.Metadata): metadata of the RampartGraph CRD
```
Comment (DianLiI) [inline src/graph/controller.py]: CR

File: src/graph/controller.py [hunk 5]
```diff
@@ -97,22 +129,42 @@ async def _sync_graphs(self):
...
+
+        # Ensure that only one task can update the state of deployment_locks at a time
         async with self._deployment_lock:
             if metadata.uid not in self._deployment_locks:
+                # Ensure that we only have one active deployment task per RampartGraph at a time
                 self._deployment_locks[metadata.uid] = asyncio.Lock()
 
         logger = GraphLogger(LOG, {"metadata": metadata})
 
+        # Fetch the RampartGraph CRD object again to make sure we are using the most recent
```
Comment (DianLiI) [inline src/graph/controller.py]: CR. Same to the line below

File: src/graph/controller.py [hunk 6]
```diff
@@ -97,22 +129,42 @@ async def _sync_graphs(self):
...
+        # field to the metadata. If the graph specification is `deploy: False`, then we also
+        # need to undeploy the graph
         if graph_obj is None or "deletionTimestamp" in graph_obj["metadata"] or \
                 not graph_obj["spec"].get("deploy", True):
             is_deletion_task = True
         else:
             cur_generation = graph_obj["metadata"]["generation"]
 
+        # The following block of code manages interrupting existing deployment tasks.
+        # For example, if the current deployment is stuck, updating/deleting the CRD
```
Comment (DianLiI) [inline src/graph/controller.py]: CR

File: src/graph/controller.py [hunk 7]
```diff
@@ -151,10 +203,14 @@ async def _sync_graph(self, metadata):
 
         async with self._deployment_locks[metadata.uid]:
             if wait_for_cancellation:
+                # task.cancel() was called before. Here we wait for that cancelation to complete
                 try:
                     await task
                 except asyncio.CancelledError:
                     pass
+            # Here we create the deployement task for the CRD,
```
Comment (DianLiI) [inline src/graph/controller.py]: CR

File: src/graph/controller.py [hunk 8]
```diff
@@ -166,6 +222,14 @@ async def _sync_graph(self, metadata):
             del self._deployment_tasks[metadata.uid]
 
     async def _sync_graph_task(self, metadata, logger):
+        """
+        This function does the actual work of updating the cluster to reflect the state
+        desired by the RampartGraph CRD.
```
Comment (DianLiI) [inline src/graph/controller.py]: CR

File: src/graph/controller.py [hunk 9]
```diff
@@ -166,6 +222,14 @@ async def _sync_graph(self, metadata):
...
+
+        args:
+            metadata (objects.base_type.Metadata): metadata of the RampartGraph CRD
```
Comment (DianLiI) [inline src/graph/controller.py]: CR

File: src/graph/controller.py [hunk 10]
```diff
@@ -176,7 +240,10 @@ async def _sync_graph_task(self, metadata, logger):
         # deleted is marked for deletion
         if "deletionTimestamp" in graph_obj["metadata"]:
             logger.info("graph marked for deletion", extra={"phase": GraphPhase.TEARDOWN})
+            # iff the finalizer exists in the CRD, then we haven't completed teardown
```
Comment (DianLiI) [inline src/graph/controller.py]: CR

File: src/graph/controller.py [hunk 11]
```diff
@@ -176,7 +240,10 @@ async def _sync_graph_task(self, metadata, logger):
...
             if FINALIZER in graph_obj["metadata"]["finalizers"]:
+                # Perform teardown and then patch the CRD to remove the finalizer
```
Comment (DianLiI) [inline src/graph/controller.py]: CR

File: src/graph/controller.py [hunk 12]
```diff
@@ -202,28 +269,41 @@ async def _sync_graph_task(self, metadata, logger):
                 except kubernetes.client.rest.ApiException:
                     return None
             return None
+        # `status.observedGeneration` is the latest value of `metadata.generation` that we
+        # have finished deployment. If `status.observedGeneration` is the same as
+        # `metadata.generation` then the current CRD has been deployed and we do not need to
```
Comment (DianLiI) [inline src/graph/controller.py]: CR

File: src/graph/controller.py [hunk 13]
```diff
@@ -233,15 +313,18 @@ async def _sync_graph_task(self, metadata, logger):
...
+                # left in a partial deployment. However, this will also cause the
+                # `status.observedGeneration` field to not be updated,
+                # so this phase will be retried.
+                #
+                # It is critical that the `graph.deploy()` implementation must finish the
+                # deployment that the previous partial deployment started.
+                # Also note that this requirement is generally met by the same functionality
+                # required by `graph.deploy()` to handle redeployment with an updated graph object.
                 try:
+                    # By adding a custom finalizer, deletion of the RampartGraph CRD
```
Comment (DianLiI) [inline src/graph/controller.py]: CR

File: src/graph/controller.py [hunk 14]
```diff
@@ -282,12 +367,19 @@ async def _sync_graph_task(self, metadata, logger):
                 else:
                     graph_obj["status"]["phase"] = "Undeployed"
 
-                # TODO: Move "endpointPaths" related stuffs to another file
+                # Deployment/Undeployment is functionally complete at this point.
+                # From here until the end is just updating the RampartGraph CRD to reflect that
```
Comment (DianLiI) [inline src/graph/controller.py]: CR

File: src/graph/controller.py [hunk 15]
```diff
@@ -308,8 +400,12 @@ async def _sync_graph_task(self, metadata, logger):
                         comp_to_paths[comp_name].update(ar_paths)
                 for key, val in comp_to_paths.items():
                     comp_to_paths[key] = list(val)
+
                 graph_obj["status"]["endpointPaths"] = json.dumps(comp_to_paths)
                 graph_obj["status"]["componentNamespaces"] = graph.namespaces
+
+                # This last `observedGeneration` value lets the controller know that it doesn't
+                # need to redeploy this version of the RampartGraph CRD
```
Comment (DianLiI) [inline src/graph/controller.py]: CR

File: src/graph/controller.py [hunk 16]
```diff
@@ -336,6 +432,7 @@ async def _sync_graph_task(self, metadata, logger):
                         raise
 
     async def _get_graph(self, metadata):
+        """Return the RampartGraph CRD object for the given metadata."""
```
Comment (DianLiI) [inline src/graph/controller.py]: CR

File: src/graph/objects/component_chart.py [hunk 17]
```diff
@@ -201,6 +218,9 @@ def _generate_repo_name(self):
 
 
 class SecretComponentChart(ComponentChart):
+    """
+    A component chart whos repo is specified by the RampartGraph CRD as a kubernetes secret.
```
Comment (DianLiI) [inline src/graph/objects/component_chart.py]: CR

File: src/graph/objects/component_chart.py [hunk 18]
```diff
@@ -164,6 +178,9 @@ def _generate_repo_name(self):
 
 
 class URLComponentChart(ComponentChart):
+    """
+    A component chart whos repo is specified by the RampartGraph CRD as a URL
```
Comment (DianLiI) [inline src/graph/objects/component_chart.py]: CR

File: src/graph/validator.py [hunk 19]
```diff
@@ -112,6 +127,16 @@ async def _validate(self, request, metadata, generated_name):
                     }
 
     async def _handle_conversion(self, request):
+        """
+        Currently there are two versions `v1beta1` and `v1beta2`. The only difference between
+        the two is that `v1beta2` adds a field `spec.deploy`: boolean, which specifies if
+        we want the graph to actually be deployed. Setting it to false will undeploy everything,
+        but leave the CRD.
```
Comment (DianLiI) [inline src/graph/validator.py]: CR

File: src/graph/objects/component.py [hunk 20]
```diff
@@ -121,6 +140,9 @@ async def validate(self, validate_specs=True):
             raise ValidationError({"'rampart-flows' is not allowed as a component name"})
 
         async def _validate_edges(target: dict, target_spec: dict):
+            """
+            Function to generalize the code for both input and output edges
```
Comment (DianLiI) [inline src/graph/objects/component.py]: What does `generalize the code` mean here?

File: src/graph/objects/component.py [hunk 21]
```diff
@@ -109,6 +116,18 @@ def needs(self):
         return self._needs
 
     async def validate(self, validate_specs=True):
+        """
+        Throws a `ValidationError` if the component is invalid. This call must be made
```
Comment (DianLiI) [inline src/graph/objects/component.py]: Raise is a more proper word for Python. Google style also has a `Raises` section like `Args` below. Same to comments under other `validate` methods.

File: src/graph/objects/component.py [hunk 22]
```diff
@@ -109,6 +116,18 @@ def needs(self):
...
+        before any function with the `@required_validated` decorator is called.
+
+        args:
```
Comment (DianLiI) [inline src/graph/objects/component.py]: Nit: It should be `Args` according to google style guide. Same to all the other docstrings.

File: src/graph/objects/component.py [hunk 23]
```diff
@@ -279,6 +329,7 @@ def _get_edge_helm_yamls(self):
 
     @require_validated
     async def _check_helm_modified(self, filename, edges_helm_name, helm_chart):
+        """Determines if upgrading would result in changes to the cluster"""
```
Comment (DianLiI) [inline src/graph/objects/component.py:339]: changes to the helm release?

File: src/graph_service/flows/run_app.py
Comment (DianLiI) [inline src/graph_service/flows/run_app.py:1]: Maybe add a module level description here.

File: src/graph_service/flows/volume_flow_liveness.py
Comment (DianLiI) [inline src/graph_service/flows/volume_flow_liveness.py:1]: A module level description will be helpful.

File: src/graph_service/run_app.py
Comment (DianLiI) [inline src/graph_service/run_app.py:1]: Maybe a module level description explicitly call out that this is the backend server for Rampart UI

File: src/graph/controller.py (hunk 1 above)
Comment (rmfan) [reply src/graph/controller.py]: I think its important to distinguish between Rampart graphs as a general, abstract concept and RampartGraph CR's as a specific kubernetes object. The CR makes it clear that its not a missing space or anything like that.

**Review by DianLiI — APPROVED** [review]

**Pull request merged**